cli/src/main/scala/scalaxb/compiler/xsd/ContextProcessor.scala

```diff
@@ -570,19 +570,27 @@ trait ContextProcessor extends ScalaNames with PackageName {
   } getOrElse {""}
       
   def identifier(value: String) = {
-    val nonspace = 
-      if (value == "") "blank" // treat "" as "blank" but " " as "u32"
-      else if (value.trim != "") """\s""".r.replaceAllIn(value, "")
-      else value    
+    def normalize(c: Char): String =
+      if (config.discardNonIdentifierCharacters) "" else s"u${c.toInt}"
+
+    // treat "" as "blank" but " " as "u32"
+    val nonspace =
+      if (value.trim != "") """\s""".r.replaceAllIn(value, "")
+      else value
     val validfirstchar =
       if ("""\W""".r.findFirstIn(nonspace).isDefined) {
           (nonspace.toSeq map { c =>
-            if ("""\W""".r.findFirstIn(c.toString).isDefined) "u" + c.toInt.toString
+            if ("""\W""".r.findFirstIn(c.toString).isDefined) normalize(c)
             else c.toString
           }).mkString
       }
       else nonspace
-    if (validfirstchar.endsWith("_")) validfirstchar.dropRight(1) + "u93"
+
+    // Scala identifiers must not end with an underscore
+    // Known issue: if `discardNonIdentifierCharacters` is set and an identifier ends in multiple underscores (e.g. `el__`)
+    //              then the generated name will be invalid (`el_`, as only the last underscore will be dropped)
+    if (validfirstchar.endsWith("_")) validfirstchar.dropRight(1) + normalize(validfirstchar.last)
+    else if (validfirstchar == "") "blank"
```

Comment:
This line had to be moved down here in case calls to normalize resulted in discarding all characters.